c++/src/kj/mutex.c++

```diff
@@ -172,6 +172,65 @@ void Mutex::lock(Exclusivity exclusivity) {
   }
 }
 
+bool Mutex::lockWithTimeout(Exclusivity exclusivity, Duration timeout) {
+  auto spec = toRelativeTimespec(timeout);
+  switch (exclusivity) {
+    case EXCLUSIVE:
+      for (;;) {
+        uint state = 0;
+        if (KJ_LIKELY(__atomic_compare_exchange_n(&futex, &state, EXCLUSIVE_HELD, false,
+                                                  __ATOMIC_ACQUIRE, __ATOMIC_RELAXED))) {
+          // Acquired.
+          break;
+        }
+
+        // The mutex is contended.  Set the exclusive-requested bit and wait.
+        if ((state & EXCLUSIVE_REQUESTED) == 0) {
+          if (!__atomic_compare_exchange_n(&futex, &state, state | EXCLUSIVE_REQUESTED, false,
+                                           __ATOMIC_RELAXED, __ATOMIC_RELAXED)) {
+            // Oops, the state changed before we could set the request bit.  Start over.
+            continue;
+          }
+
+          state |= EXCLUSIVE_REQUESTED;
+        }
+
+        auto result = syscall(SYS_futex, &futex, FUTEX_WAIT_PRIVATE, state, &spec, nullptr, 0);
+        if (result < 0) {
+          // We timed out, remove exclusive requested flag.
+          if (errno == ETIMEDOUT) {
+            __atomic_fetch_and(&futex, ~EXCLUSIVE_REQUESTED, __ATOMIC_RELEASE);
+            return false;
+          }
+        }
+      }
+      break;
+    case SHARED: {
+      uint state = __atomic_add_fetch(&futex, 1, __ATOMIC_ACQUIRE);
+      for (;;) {
+        if (KJ_LIKELY((state & EXCLUSIVE_HELD) == 0)) {
+          // Acquired.
+          break;
+        }
+
+        // The mutex is exclusively locked by another thread.  Since we incremented the counter
+        // already, we just have to wait for it to be unlocked.
+        auto result = syscall(SYS_futex, &futex, FUTEX_WAIT_PRIVATE, state, &spec, nullptr, 0);
+        if (result < 0) {
+          // If we timeout though, we need to signal that we're not waiting anymore.
+          if (errno == ETIMEDOUT) {
+            state = __atomic_sub_fetch(&futex, 1, __ATOMIC_ACQUIRE);
```

Comment:
Is this the correct ordering? I'm still not 100% sure on what orderings should be used when.

Reply:
You can use `__ATOMIC_RELAXED` here, because the calling thread hasn't touched and will not touch the protected object. Generally, `__ATOMIC_ACQUIRE` is needed if the calling thread will be accessing the protected object after the call (but was not permitted to do so before), and `__ATOMIC_RELEASE` is needed if it accessed the protected object before the call (but won't be permitted to after).